internal/graph/check.go

```diff
@@ -572,62 +572,182 @@ func checkAssociatedObjects(ctx context.Context, req *ResolveCheckRequest, objec
+	defer close(dispatches)
+	reqTupleKey := req.GetTupleKey()
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
 	for {
 		t, err := iter.Next(ctx)
 		if err != nil {
-			if errors.Is(err, storage.ErrIteratorDone) {
+			// cancelled doesn't need to flush nor send errors back to main routine
+			if storage.IterIsDoneOrCancelled(err) {
 				break
 			}
-			telemetry.TraceError(span, err)
-			return nil, err
+			trySendDispatchChan(ctx, dispatchMsg{err: err}, dispatches)
+			break
 		}
 
 		usersetObject, usersetRelation := tuple.SplitObjectRelation(t.GetUser())
-		reqTupleKey := req.GetTupleKey()
 
 		// if the user value is a typed wildcard and the type of the wildcard
 		// matches the target user objectType, then we're done searching
 		if tuple.IsTypedWildcard(usersetObject) && typesystem.IsSchemaVersionSupported(typesys.GetSchemaVersion()) {
 			wildcardType := tuple.GetType(usersetObject)
 
 			if tuple.GetType(reqTupleKey.GetUser()) == wildcardType {
-				span.SetAttributes(attribute.Bool("allowed", true))
-				response.Allowed = true
-				return response, nil
+				trySendDispatchChan(ctx, dispatchMsg{shortCircuit: true}, dispatches)
+				break
 			}
-
-			continue
 		}
 
 		if usersetRelation != "" {
 			tupleKey := tuple.NewTupleKey(usersetObject, usersetRelation, reqTupleKey.GetUser())
-			handlers = append(handlers, c.dispatch(ctx, req, tupleKey))
+			trySendDispatchChan(ctx, dispatchMsg{dispatchParams: &dispatchParams{parentReq: req, tk: tupleKey}}, dispatches)
 		}
 	}
+}
 
-	resp, err := union(ctx, c.concurrencyLimit, handlers...)
+func (c *LocalChecker) processDispatches(ctx context.Context, pool *pool.ContextPool, dispatchChan chan dispatchMsg, outcomes chan checkOutcome) {
+	defer func() {
+		// We need to wait always to avoid a goroutine leak.
+		_ = pool.Wait()
+		close(outcomes)
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case msg, ok := <-dispatchChan:
+			if !ok {
+				return
+			}
+			if msg.err != nil {
+				outcomes <- checkOutcome{err: msg.err}
+				break // continue
+			}
+			if msg.shortCircuit {
+				outcomes <- checkOutcome{resp: &ResolveCheckResponse{
+					Allowed: true,
+					ResolutionMetadata: &ResolveCheckResponseMetadata{
+						DatastoreQueryCount: 0,
+					},
+				}}
+				return
+			}
+
+			if msg.dispatchParams != nil {
+				pool.Go(func(ctx context.Context) error {
+					resp, err := c.dispatch(ctx, msg.dispatchParams.parentReq, msg.dispatchParams.tk)(ctx)
+					outcomes <- checkOutcome{resp: resp, err: err}
+					return nil
+				})
+			}
+		}
+	}
+}
+
+func (c *LocalChecker) consumeDispatches(ctx context.Context, req *ResolveCheckRequest, limit uint32, dispatchChan chan dispatchMsg) (*ResolveCheckResponse, error) {
+	outcomes := make(chan checkOutcome, limit)
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	dispatchPool := concurrency.NewPool(cancellableCtx, int(limit))
+
+	go c.processDispatches(ctx, dispatchPool, dispatchChan, outcomes)
```

Comment:
Notice two things:
- `dispatchPool := concurrency.NewPool(cancellableCtx, int(limit))` is passed as parameter and then never used again in this function - IMO, code smell that this really belongs inside `processDispatches`
- `processDispatches` doesn't return anything, which makes it slightly harder to unit test

We can improve `processDispatches` by having it return something: the channel. I think this is better than having to construct it outside and pass it in (because you will also have to do that in the tests).

The calling code would be

```go
func (c *LocalChecker) consumeDispatches(ctx context.Context, req *ResolveCheckRequest, limit uint32, dispatchChan chan dispatchMsg) (*ResolveCheckResponse, error) {
	cancellableCtx, cancel := context.WithCancel(ctx)
	outcomeChannel := c.processDispatches(cancellableCtx, limit, dispatchChan)
```

and then `processDispatches` would be 


```go
func (c *LocalChecker) processDispatches(ctx context.Context, limit uint32, dispatchChan chan dispatchMsg) chan checkOutcome {
	outcomes := make(chan checkOutcome, limit)
	dispatchPool := concurrency.NewPool(ctx, int(limit))

	go func() {
		// existing code (including defer)
	}()

	return outcomes
}
```

This is the proposal: [1888sug.patch](https://github.com/user-attachments/files/16912367/1888sug.patch). It passes tests locally :)